[SPARK-45894][SQL] hive table level setting hadoop.mapred.max.split.size

guixiaowen opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
In the scenario of hive table scan, by configuring the hadoop.mapred.max.split.size parameter, you can increase the parallelism of the scan hive table stage, thereby reducing the running time.

However, if a large table and a small table are in the same query, if only a separate hadoop.mapred.max.split.size parameter is configured, some stages will run a very large number of tasks, and some stages will The number of tasks running is very small. For runtime tasks, the hadoop.mapred.max.split.size parameter can be set separately for each hive table to ensure this balance.




<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->


### Why are the changes needed?
If a large table and a small table are in the same query, if only a separate hadoop.mapred.max.split.size parameter is configured, some stages will run a very large number of tasks, and some stages will The number of tasks running is very small. 

<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->


### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->

guixiaowen committed:
SPARK-45894 hive table level setting hadoop.mapred.max.split.size

dongjoon-hyun reviewed (changes requested):
Thank you for making a PR, @guixiaowen . I have a few comments.

1. Apache Spark 3.2.0 is the end of life as of now. So, this improvement should have 4.0.0.
```
- .version("3.2.0")
+ .version("4.0.0")
```

2. Given that this is a Hive-table-specific configuration and will not change during the lifetime of Spark jobs, please move the config to `HiveUtils.scala` like the following.

https://github.com/apache/spark/blob/9987cab84bb05a61a5c0b43f94a733561a2e074a/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala#L157

3. The config name, `spark.hadoop.mapred.`, looks a little weird because `org.apache.hadoop.mapred` is the older API and `org.apache.hadoop.mapreduce` is the new one. In addition, this PR also uses `mapreduce.input.*` in the code.

4. We are careful about adding a new namespace in the new config addition. For example, `spark.hadoop.mapred.max.split.size.by.table`, is creating multiple namespaces like the following. If there is no other sibiling in the same namespace, we should not use `.`.
```
spark.hadoop.mapred.*
spark.hadoop.mapred.max.*
spark.hadoop.mapred.max.split.*
spark.hadoop.mapred.max.split.size.*
spark.hadoop.mapred.max.split.size.by.*
```

5. I'm wondering if this is a general improvement. Specifically, Hive Table property is helpful your case or not? For example, IIUC, Spark understands `TBLPROPERTIES ("orc.compress"="ZLIB")` style. Do you need to put your `mapreduce.input.fileinputformat.split.maxsize` into Hive Table property in the same way instead of adding a new Spsark configuration?

6. We need a test coverage to protect your PR from the possible future regressions.

guixiaowen commented:
> Thank you for making a PR, @guixiaowen . I have a few comments.
> 
> 1. Apache Spark 3.2.0 is the end of life as of now. So, this improvement should have 4.0.0.
> 
> ```
> - .version("3.2.0")
> + .version("4.0.0")
> ```
> 
> 2. Given that this is a Hive-table-specific configuration and will not change during the lifetime of Spark jobs, please move the config to `HiveUtils.scala` like the following.
> 
> https://github.com/apache/spark/blob/9987cab84bb05a61a5c0b43f94a733561a2e074a/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala#L157
> 
> 3. The config name, `spark.hadoop.mapred.`, looks a little weird because `org.apache.hadoop.mapred` is the older API and `org.apache.hadoop.mapreduce` is the new one. In addition, this PR also uses `mapreduce.input.*` in the code.
> 4. We are careful about adding a new namespace in the new config addition. For example, `spark.hadoop.mapred.max.split.size.by.table`, is creating multiple namespaces like the following. If there is no other sibiling in the same namespace, we should not use `.`.
> 
> ```
> spark.hadoop.mapred.*
> spark.hadoop.mapred.max.*
> spark.hadoop.mapred.max.split.*
> spark.hadoop.mapred.max.split.size.*
> spark.hadoop.mapred.max.split.size.by.*
> ```
> 
> 5. I'm wondering if this is a general improvement. Specifically, Hive Table property is helpful your case or not? For example, IIUC, Spark understands `TBLPROPERTIES ("orc.compress"="ZLIB")` style. Do you need to put your `mapreduce.input.fileinputformat.split.maxsize` into Hive Table property in the same way instead of adding a new Spsark configuration?
> 6. We need a test coverage to protect your PR from the possible future regressions.

Thank you very much for viewing.
I will modify it according to the points you mentioned.

guixiaowen committed:
[SPARK-45894][SQL] hive table level setting hadoop.mapred.max.split.size #43768
